packages/@aws-cdk/aws-cognito/lib/user-pool-idp.ts

```diff
@@ -29,4 +32,388 @@ export class UserPoolIdentityProvider {
+   */
+  public static other(attributeName: string): ProviderAttribute {
+    return new ProviderAttribute(attributeName);
+  }
+
+  /** The attribute value string as recognized by the provider. */
+  public readonly attributeName: string;
+
+  private constructor(attributeName: string) {
+    this.attributeName = attributeName;
+  }
+}
+
+/**
+ * The mapping of user pool attributes to the attributes provided by the identity providers.
+ */
+export interface AttributeMapping {
+  /**
+   * The user's postal address is a required attribute.
+   * @default - not mapped
+   */
+  readonly address?: ProviderAttribute;
+
+  /**
+   * The user's birthday.
+   * @default - not mapped
+   */
+  readonly birthdate?: ProviderAttribute;
+
+  /**
+   * The user's e-mail address.
+   * @default - not mapped
+   */
+  readonly email?: ProviderAttribute;
+
+  /**
+   * The surname or last name of user.
+   * @default - not mapped
+   */
+  readonly familyName?: ProviderAttribute;
+
+  /**
+   * The user's gender.
+   * @default - not mapped
+   */
+  readonly gender?: ProviderAttribute;
+
+  /**
+   * The user's first name or give name.
+   * @default - not mapped
+   */
+  readonly givenName?: ProviderAttribute;
+
+  /**
+   * The user's locale.
+   * @default - not mapped
+   */
+  readonly locale?: ProviderAttribute;
+
+  /**
+   * The user's middle name.
+   * @default - not mapped
+   */
+  readonly middleName?: ProviderAttribute;
+
+  /**
+   * The user's full name in displayable form.
+   * @default - not mapped
+   */
+  readonly fullname?: ProviderAttribute;
+
+  /**
+   * The user's nickname or casual name.
+   * @default - not mapped
+   */
+  readonly nickname?: ProviderAttribute;
+
+  /**
+   * The user's telephone number.
+   * @default - not mapped
+   */
+  readonly phoneNumber?: ProviderAttribute;
+
+  /**
+   * The URL to the user's profile picture.
+   * @default - not mapped
+   */
+  readonly profilePicture?: ProviderAttribute;
+
+  /**
+   * The user's preferred username.
+   * @default - not mapped
+   */
+  readonly preferredUsername?: ProviderAttribute;
+
+  /**
+   * The URL to the user's profile page.
+   * @default - not mapped
+   */
+  readonly profilePage?: ProviderAttribute;
+
+  /**
+   * The user's time zone.
+   * @default - not mapped
+   */
+  readonly timezone?: ProviderAttribute;
+
+  /**
+   * Time, the user's information was last updated.
+   * @default - not mapped
+   */
+  readonly lastUpdateTime?: ProviderAttribute;
+
+  /**
+   * The URL to the user's web page or blog.
+   * @default - not mapped
+   */
+  readonly website?: ProviderAttribute;
+
+  /**
+   * Specify custom attribute mapping here and mapping for any standard attributes not supported yet.
+   * @default - no custom attribute mapping
+   */
+  readonly custom?: { [key: string]: ProviderAttribute };
+}
+
+/**
+ * Properties to create a new instance of UserPoolIdentityProvider
+ */
+export interface UserPoolIdentityProviderProps {
```

Comment:
Can this also be non-exported?

Reply:
> How about instead moving UserPoolIdentifyProviderBase off into a separate file, maybe under `lib/private/`? How complicated would that look?
> I like the separate file structure we already have and want to see if we can maintain that.

this is actually what I initially tried (mentioned it in the commit body) but JSII will not allow it. I also tried to preserve the structure initially.

as for the props,  we need to export it as it's an extended interface. All our constructs export the base interface. (jsii also throws an error if you exclude it)

Reply:
Ok, that's fine.

Can we open issues in JSII for these two? I'd like to hear from the JSII team if this is unsupported for a reason or just a gap.

Reply:
will do!